neonvm/controllers/metrics.go

```diff
@@ -176,28 +182,27 @@ func (d *wrappedReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 			// if a VM is getting conflict errors, it doesn't mean
 			// non-conflict errors are resolved, as they are more
 			// likely to be persistent.
-			delete(d.conflicting, req.NamespacedName)
+			d.conflicting.RecordSuccess(req.NamespacedName)
 		}
 
 		log.Error(err, "Failed to reconcile VirtualMachine",
 			"duration", duration.String(), "outcome", outcome)
 	} else {
-		delete(d.failing, req.NamespacedName)
-		delete(d.conflicting, req.NamespacedName)
+		d.failing.RecordSuccess(req.NamespacedName)
 		log.Info("Successful reconciliation", "duration", duration.String())
 	}
 	d.Metrics.ObserveReconcileDuration(outcome, duration)
 	d.Metrics.failing.WithLabelValues(d.ControllerName,
-		string(FailureOutcome)).Set(float64(len(d.failing)))
+		string(FailureOutcome)).Set(float64(d.failing.FiringCount()))
```

Comment:
Technically, things can become firing outside of reconcile cycles, right?

It probably doesn't matter in practice, but should we instead just have a periodic job to update the metrics (rather than updating it every time?)

Reply:
That's right. I guess I could spawn another goroutine, which would fetch the counter and put it into `GaugeVec`. This would mean to move [that](https://github.com/neondatabase/autoscaling/pull/757/files#diff-2d9cf99104a80ac4cc6080fbe379163550a313e7e3734c631fb896f8f4630ed4R55) mutex into the `Tracker`, and make `Tracker` thread-safe, but this means mutex is on a hotpath. 

I don't expect any significant contention around mutex wrapping a map, not sure why Shayan was worried. 

Another approach would be to have a [GaugeFunc](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#GaugeFunc) instead of a goroutine, but it is a little inconvenient, as there is no `GaugeVecFunc`. 

Thoughts?

Reply:
Hm... there's a few options here, huh.

Acquiring many mutexes per reconcile instead of a single one might not be great, although we'd have to see in practice whether it affects reconcile latency (might be good to test out this PR on staging?). Off-hand, I suspect it's _probably_ fine.

I've hit the "no GaugeVecFunc" issue before... its unfortunate. Could write our own GaugeVecFunc equivalent, but that's probably a lot of work.

One related thought:

For the "stuck VMs" dashboard, it's helpful that the autoscaler-agent periodically logs that the VM is marked "stuck" (that way, we can see which VMs were having issues after-the-fact -- without requiring some debug endpoint to fetch the current list). What if we did something similar here? Idea would be:

> Periodically:
> 1. Fetch all firing VMs
> 2. Update the metrics
> 3. Log which ones are firing

Could have a new config parameter `RefreshFailingInterval` ;)